### DISCUSSION TITLE
Need of coercing array_like input data?

### DISCUSSION THREAD

**Discussion by jcjaskula-aws** [lmfit/lmfit-py#873] [Q&A]
Hello,

Input data (tuples, lists and pd.Series) is now transformed into numpy.array and returned back in their new type in `ModelResult.data`. Recasting `pandas.Series` in `numpy.array` results in a loss of convenient functionalities. 

```
y=pd.Series(range(8), pd.MultiIndex.from_product([[1,1,2,3], ["a", "b"]], names=["digits", "letters"]))
m=lmfit.models.LinearModel().fit(y, x=[1,1,1,1,2,2,3,3])
m.eval(x=m.data.index.unique(level="digits"))
```

This was not the case before since this [commit](https://github.com/lmfit/lmfit-py/commit/25d0a85a0b767c45453ca737395567a196807491) when it was very convenient to use pandas functionalities directly on the fitted data. 

What's the reason behind this? Is keeping this constraint mandatory?

**Comment by newville** [comment]
Post a minimal, complete example that demonstrates what you are seeing, and be explicit about what you expected to see and what you actually see.  Am I supposed to know what you think is wrong?  I do not.

Yes, some things changed with 1.2.0 for how data is coerced into numpy arrays.     

I think it is not correct that we change the data you passed in: that is, your `y` should not be changed and still be your `Series`. 

Internally, data and weights for a model are converted to numpy arrays of type float64.  Independent data is of type list, tuple, pandas.Series can easily be interpreted as "really meant to be a float64 array" and is converted.    Independent data of other data types, especially data that is ndarrays of different precision, or dicts, are not converted.

**Comment by jcjaskula-aws** [comment]
I agree with your statement that data passed in should not be changed. My post here is to report my experience that they are modified and I would like to change this behavior.

Let me simplify and be more explicit.
```
import pandas as pd
import lmfit
y=pd.Series(range(8), pd.MultiIndex.from_product([[1,1,2,3], ["a", "b"]], names=["digits", "letters"]))
y
digits  letters
1       a          0
        b          1
        a          2
        b          3
2       a          4
        b          5
3       a          6
        b          7
dtype: int64
```
is a pandas.Series with multiple indices.  If I run `y.index.unique(level="digits")`, I get a list of unique first indices: `Int64Index([1, 2, 3], dtype='int64', name='digits')`.

if I understand correctly, when I run
```
m=lmfit.models.LinearModel().fit(y, x=[1,1,1,1,2,2,3,3])
```
`y` and `m.data` should the the same data.

However, I get the following error with `m.data.index.unique(level="digits")`:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 3
      1 y=pd.Series(range(8), pd.MultiIndex.from_product([[1,1,2,3], ["a", "b"]], names=["digits", "letters"]))
      2 m=lmfit.models.LinearModel().fit(y, x=[1,1,1,1,2,2,3,3])
----> 3 m.eval(x=m.data.index.unique(level="digits"))

AttributeError: 'numpy.ndarray' object has no attribute 'index'
```
This is just a complicated way to say that `type(m.data)` is different from `type(y)`.
I would advocate to insure that `ModelResult.data` is not modified, even though lmfit needs an internal numpy.array copy. 

MVE:
```
import pandas as pd
import lmfit
y=pd.Series(range(8))
m=lmfit.models.LinearModel().fit(y, x=list(range(8)))
assert type(y) == type(m.data)

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[7], line 5
      3 y=pd.Series(range(8))
      4 m=lmfit.models.LinearModel().fit(y, x=list(range(8)))
----> 5 assert type(y) == type(m.data)

AssertionError: 
```

**Reply by newville** [reply]
OK, good.  That is convincing that your `y` is not changed.   That would be a problem.   

The fact that `m.data` is a ndarray with `dtype` of `float64` is by design.

> if I understand correctly, when I run
>    m=lmfit.models.LinearModel().fit(y, x=[1,1,1,1,2,2,3,3])
> y and m.data should the the same data.

`m.data` is used internally and may be coerced to a ndarray of datatype `float64` or `complex128`.

**Reply by jcjaskula-aws** [reply]
Then, I would advocate for using a private variable `m._data` when it is used internally. `ModelResult.data` can be used to vend back the input data and it is convenient to keep the pandas type for instance, which is what I tried to motivate above.

**Reply by newville** [reply]
@jcjaskula-aws I don't understand what renaming `Model.data` to `Model._data` would accomplish.   You have a pandas Series and want to fit the data in that Series.  That should work.  I think you have not shown that it does not work.  What you showed was that `Model.data` is an ndarray.  That seems correct to me.

**Reply by jcjaskula-aws** [reply]
Sorry, my point would be to use `ModelResult._data` internally as `numpy.array` and `ModelResult.data` exactly as the user input data. To give more context, I'm still using an old version dated of May 2021 where `ModelResult.data` is not cast into ndarray and I process this data further after fitting for plotting purposes. I cannot do it the same way with newer version of lmfit and I would need to change my code to carry the `Series` along with `ModelResults`.

**Reply by reneeotten** [reply]
I agree with @newville that this change was on purpose and is not likely to be reverted... I suppose that you still have your `y` values available so why not use those instead of using `ModelResult.data`? If the point is that you'd like to be able to Save/Load a ModelResult later and still use the `pandas.Series` then I can see your point.

I just remembered there is an attribute `user_options` in the `ModelResult` class, which I think you could use to put your `pandas.Series` in so that it can be saved/loaded. I don't see an obvious other use for it nor a very clear way of passing it to this attribute except for just assigning it after the fit. Perhaps @newville remembers the exact use-case for this attribute and/or whether that would make sense to (ab)use here ;)

**Reply by newville** [reply]
@jcjaskula-aws Model takes the result of running the model function (`model`) and the data (`data`) and possibly a weighting factor (`weights`) and it calculates a residual array of `(data - model)*weights`.  That resulting array must be a ndarray with datatype of `float64`.   Just to be clear, that is not "should" and it is not "must be coercible to float64", it really is "must" in the sense that the fitting algorithm will abort immediately if it does not get a float64 ndarray.  Pandas Series are not float64 ndarrays, neither are float32 ndarrays. 

Lmfit needs to calculate that residual array after each evaluation of the model function - we assume that could be many times.   The return value from the model function is not something we have a lot of control over.    The parameter values passed in will be Python floats, but the independent data can be whatever the user wants (int8 arrays, dicts, pandas Series, other objects, etc) and the user's model function can basically do whatever it wants with those and sends something back.  

We need `data-model` to be float64 ndarray for every function evaluation.   Coercing `data` to float64 ndarray before the function evaluations goes a long way to ensure that will be the case.  And, we just do not care at all if the user passed in a pandas Series for `data` -- lmfit is only going to use that as a float64 ndarray. 

I think it would be completely defensible for `Model` (and other parts of `lmfit`) to raise a ValueError if `data` (or `weights`, or the return value of the Model function) was not strictly a float64 ndarray.    But if the user sends in something (list, pandas.Series, ndarray with different datatype) that is easily coerced to float64 ndarray, there is no harm in us just doing that.   But there is no need for us to carry around a pandas.Series - that is not what is used in the fit.

If you want some other aspect of a Series saved (though I still do not follow your use case), then perhaps you want to save those parts separately or pass in that Series as independent data as well as the primary data.

**Reply by jcjaskula-aws** [reply]
Thanks both of you for these explanations, that's what I was looking for. 
lmfit is deeply integrated in my software and leverage the fact that `ModelResult.data` is an array-like object (according to the docstring and my current experience with 1.0.2 that is old I recognize). I was maybe abusing this attribute to carry my data through my program. I agree that there is nothing fundamentally blocking me and I could implement a way to pass `y` explicitly. The main problem that I raised here is that this coercing change is effectively an API change and eventually prevents me to upgrade lmfit straightforwardly. If anything, we should update the [docstring](https://github.com/lmfit/lmfit-py/blob/a246cbc7daa1190b774d716d192b004672acdbc4/lmfit/model.py#L1372).

**Reply by newville** [reply]
@jcjaskula-aws  Well, `ModelResult.data` is array-like.  It is not Series-like.   That is what the documentation says.
 
The changes we made are not API changes.  The call structures for the methods were not changed in a backward-incompatible way.  The internals of what `Model` does with the inputs did change.  That can happen on every change and is not an API change.

It seems perfectly reasonable to me to say that trying to support `pandas.Series` has proven to be more trouble than it is worth, as the pandas users then expect that their data structure is driving the code is or is fully supported.   There are other "array-like" structures, such as `xarray` and `dask` that we do not test.  Though it might seem obvious how to convert these (or `Series`) into the ndarray we actually use, it is also reasonable to "refuse to guess" and to not automatically convert any data to ndarrays, but to raise a TypeError exception if the input data is not a float64 ndarray.    Honestly, I do not think that is necessary, but this discussion topic is definitely leading me in that direction.

**Reply by rayosborn** [reply]
Since I am not a maintainer of `lmfit`, I would certainly defer to the judgement of @newville and @reneeotten about what they should and should not support. However, I think what @jcjaskula-aws asks for could have been accomplished by using the `np.array` function on the `y` argument instead of coercing it directly. 
```
>>>  y=pd.Series(range(8), pd.MultiIndex.from_product([[1,1,2,3], ["a", "b"]], names=["digits", "letters"]))
>>> np.array(y).astype(np.float64)
array([0., 1., 2., ..., 5., 6., 7.])
```
I have checked that this also works for `xarray`, `dask`, and `NXfield` objects, the latter defined by the [nexusformat](https://nexpy.github.io/nexpy/) package (which is why I am interested). The only requirement is that the array format has to have a `__array__` function, which NumPy uses to return a valid `ndarray`, and clearly this is fairly routinely implemented if these packages are anything to go by. As @jcjaskula-aws suggests, this would need a private coerced copy of the `y` data within `lmfit`, but it doesn't involve any guess-work about what these different array formats do.

I can see arguments for not doing this, since it could add to future maintainability problems, and sorry for not suggesting this when the original data coercion debate was underway. In the case of `nexusformat`, it is just as simple to pass the `NXfield.nxvalue` property to `lmfit` to get the same result, so I'm not too bothered either way.

Edit: I should probably have checked the `lmfit` code before commenting. Perhaps you already use the `np.array` function to coerce the input, in which case the only debate is whether you should keep the coerced values in a private attribute.

**Reply by newville** [reply]
@rayosborn Yes we use `np.asfarray(x)` on the object passed in as data.   I think the only real change is that we do it early on the `data` (and `weights`) array, and then on the result of calling the model function.  FWIW, I think the main driving issue was for data sent in that was Int32. 

I think the question is whether we need to hold onto that Series, xarray, HDF5 group, etc that the user passed in or just the float64 array.

**Reply by jcjaskula-aws** [reply]
Yes, indeed I think it just boils down to how to vend back the modeled data in the model results. My point was also not to focus on Series itself but any type that is cast into ndarray (i.e. tuple, list, Series). And I'm currently wondering what would happen if I provide an array_like object that is not one of these three mentioned before (I'm not familiar with `nxvalue` but I feel it would end up in this situation). My guess (when I read [this part of the code](https://github.com/lmfit/lmfit-py/blob/502cc9ba448eb90aec832c94063a07c47c92bcf5/lmfit/model.py#L1068)) is that the type would not be coerced.

**Reply by newville** [reply]
@jcjaskula-aws My view on this might  be evolving, partly informed from other discussions. I think the recent changes are probably in the right direction.   

We definitely need to be passing a 1d Float64 ndarray to the fitting methods. It is OK if that is Complex128 ndarray or a single Float64 value for the scalar Minimizers.  I will call this float64 array in general. Other ndarray dtypes or other array-like objects must be converted to float64.  I do mean "must".

We could coercion "early" or "late", but we must do it.

For independent data used by the user to calculate the model,  I think we should never do anything to the data.  Currently we do some coercion for some cases.  

We have code throughout `model.py` that handles pandas Series specially.  I think we should probably remove all of this so that any array-like object is handled. 

Unlike the independent data, the primary data and weights need to be array-like.
When we do `(data-model)`, we currently coerce the data early and expect but do not ensure that the model is a Float64 array.  Separately,  I think we should ensure that `Model.eval` gives a Float64 
ndarray.  At that point, I think that we do not need to coerce the data,  but may want to also coerce the result `(data-model)*weights`. That is, we should probably change to "coerce late." It might add a little overhead, but that should be acceptable. 

I think that would then better allow other array-like objects without us having to add special code.  I think the main complication would be handling Nans, but again, I think that doing this late would  be OK.

**Comment by newville** [comment]
an update on this.  I just merged #899 into the master branch which addresses this issue, I think in a better way. 

First data returned by an objective function is better coerced into Float64  or Complex128 1D-array, ready to send on to the solver code.  That should make objective functions that do use "array-like" objects more guaranteed to work well.

Second, for `Model.fit()`, there are just too many examples of using plain lists floating around to actually do absolutely nothing with input data (or independent data).   

With #899, the default behavior is to "coerce early"  data and independent data that is "array-like" (list, tuple, pandas Series, h5py Datasets, anything with an `__array__()` method).  

But: the user can pass `coerce_farray=False` to `Model.fit()` which will then not do any coercion at all.  With that option, the user is able to send pandas Series, etc into their model function, but they are also then taking responsibility for working with the data types they have.     

Independent data can still be any complex object: if it is not "array-like", it is not converted to a float64 array.  But that also implies the user intended to do this.  

I think that adding  `coerce_farray=True` to `Model.fit()` strikes a decent balance between making the common cases easy (sure, the novice can send in an array of numbers or a pandas Series), while also making more complex cases possible and not too hard.
